File: sdks/python/src/verdocs/resources/signatures.py
"""Signature block operations (js-sdk: Envelopes/Signatures.ts).

The multipart plumbing lives in envelopes.py until the endpoint itself grows
files support; we share it rather than repeat the error translation here.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from ..models.core import Signature
from ..models.envelopes import FileInput
from .envelopes import _async_multipart_request, _file_part, _multipart_request

if TYPE_CHECKING:
    from .._endpoint import AsyncVerdocsEndpoint, VerdocsEndpoint

_SIGNATURES_PATH = "/v2/profiles/signatures"


class Signatures:
    """Signature block calls for a sync endpoint."""

    def __init__(self, endpoint: VerdocsEndpoint) -> None:
        self._endpoint = endpoint

    def create(self, image: FileInput) -> Signature:
        """Store a signature block image for the caller's profile.

        Sends multipart/form-data to POST /v2/profiles/signatures with the
        image under a part named "signature". Works with a user or signing
        session (guest signer profiles are created automatically). Signers
        typically "adopt" one block at the start of signing and reuse its ID
        for every signature field via envelopes.update_field. Guest signers
        usually have exactly one, tied to their session; authenticated users
        can keep several. Mirrors js-sdk createSignature, folding its
        separate name argument into the file input (the server does not
        consume the filename).

        Example:
            block = endpoint.signatures.create("/tmp/signature.png")
            endpoint.envelopes.update_field(envelope_id, "Recipient 1", "sig-1", block.id)

        Args:
            image: The signature image: a file path, raw bytes, a binary
                file-like object, or an httpx-style (filename, content,
                content_type) tuple.

        Returns:
            The stored signature block. Its url field is the storage object
            key, not a fetchable URL; the id is what signature fields accept.

        Raises:
            VerdocsAPIError: The API rejected the upload.
            VerdocsConnectionError: The request never reached the API.

        @sdkOperation signature.createSignature
        @sdkGroup Signature
        @sdkPage Endpoints
        """
        response = _multipart_request(
            self._endpoint, "POST", _SIGNATURES_PATH, files=[("signature", _file_part(image))]
        )
        return Signature.model_validate(response.json())


class AsyncSignatures:
    """Signature block calls for an async endpoint."""

    def __init__(self, endpoint: AsyncVerdocsEndpoint) -> None:
        self._endpoint = endpoint

    async def create(self, image: FileInput) -> Signature:
        """Store a signature block image for the caller's profile.

        Sends multipart/form-data to POST /v2/profiles/signatures with the
        image under a part named "signature". Works with a user or signing
        session (guest signer profiles are created automatically). Signers
        typically "adopt" one block at the start of signing and reuse its ID
        for every signature field via envelopes.update_field. Guest signers
        usually have exactly one, tied to their session; authenticated users
        can keep several. Mirrors js-sdk createSignature, folding its
        separate name argument into the file input (the server does not
        consume the filename).

        Example:
            block = await endpoint.signatures.create("/tmp/signature.png")
            await endpoint.envelopes.update_field(envelope_id, "Recipient 1", "sig-1", block.id)

        Args:
            image: The signature image: a file path, raw bytes, a binary
                file-like object, or an httpx-style (filename, content,
                content_type) tuple.

        Returns:
            The stored signature block. Its url field is the storage object
            key, not a fetchable URL; the id is what signature fields accept.

        Raises:
            VerdocsAPIError: The API rejected the upload.
            VerdocsConnectionError: The request never reached the API.
        """
        response = await _async_multipart_request(
            self._endpoint, "POST", _SIGNATURES_PATH, files=[("signature", _file_part(image))]
        )
        return Signature.model_validate(response.json())

File: sdks/csharp/src/Verdocs.Sdk/Resources/Signatures.cs
using System.Net.Http.Headers;
using Verdocs.Models;

namespace Verdocs.Resources;

/// <summary>
/// Signature block calls, reached through the endpoint's Signatures property.
/// </summary>
public sealed class Signatures
{
    private readonly VerdocsEndpoint _endpoint;

    internal Signatures(VerdocsEndpoint endpoint)
    {
        _endpoint = endpoint;
    }

    /// <summary>
    /// Creates a signature block from an image. Signers typically adopt one signature block at
    /// the start of signing, then reuse its ID for every signature field to be stamped (via
    /// the envelope field update). Accepts a user or a signing session: guest signers get an
    /// auto-created profile and typically hold one block tied to that session, while
    /// authenticated users can keep several and use them interchangeably. The stored image is
    /// not fetchable through the v2 API; the block's ID is what matters.
    /// </summary>
    /// <param name="content">The signature image to store.</param>
    /// <param name="fileName">The filename to store with the image.</param>
    /// <param name="contentType">The image's MIME type. Defaults to application/octet-stream; the server stores the declared type without validating it.</param>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>The newly created signature block.</returns>
    /// <exception cref="VerdocsApiException">The call failed, for example because the session is invalid.</exception>
    /// <sdkOperation>signature.createSignature</sdkOperation>
    /// <sdkGroup>Signature</sdkGroup>
    /// <sdkPage>Endpoints</sdkPage>
    public Task<Signature> CreateAsync(
        Stream content,
        string fileName,
        string? contentType = null,
        CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(content);
        ArgumentException.ThrowIfNullOrEmpty(fileName);

        var form = new MultipartFormDataContent();
        var part = new StreamContent(content);
        part.Headers.ContentType = new MediaTypeHeaderValue(contentType ?? "application/octet-stream");
        // Browsers always quote part names and filenames while .NET leaves token-safe values
        // unquoted, so we pre-quote to keep the wire shape identical to the js-sdk's.
        part.Headers.ContentDisposition = new ContentDispositionHeaderValue("form-data")
        {
            Name = "\"signature\"",
            FileName = "\"" + fileName + "\"",
        };
        form.Add(part);

        return _endpoint.SendAsync<Signature>(HttpMethod.Post, "/v2/profiles/signatures", form, cancellationToken);
    }
}
